[Async Refactoring] Split ambiguous (error + success) completion handler calls into success and error case

ahoppen opened the pull request:
Previously, when a completion handler call had arguments for both the success and error parameters, we were always interpreting it as an error call. In case the error argument was an `Optional`, this could cause us to generate code that didn't compile, because we `throw`ed the `Error?` or passed it to `continuation.resume(throwing)`, both of which didn't work.

We now generate an `if let` statement that checks if the error is `nil`. If it is, the error is thrown. Otherwise, we interpret the call as a success call and return the result.

## Example 1 (convert to continuation)
### Base Code
```swift
func test(completionHandler: (Int?, Error?) -> Void) {
  withoutAsyncAlternativeThrowing { (theValue, theError) in
    completionHandler(theValue, theError)
  }
}
```
### Old Refactoring Result
```swift
func test() async throws -> Int {
  return try await withCheckedThrowingContinuation { continuation in
    withoutAsyncAlternativeThrowing { (theValue, theError) in
      continuation.resume(throwing: theError) // error: Argument type 'Error?' does not conform to expected type 'Error'
    }
  }
}
```
### New Refactoring Result
```swift
func testThrowingContinuationRelayingErrorAndResult() async throws -> Int {
  return try await withCheckedThrowingContinuation { continuation in 
    withoutAsyncAlternativeThrowing { (theValue, theError) in
      if let error = theError {
        continuation.resume(throwing: error)
      } else {
        guard let theValue = theValue else {
          fatalError("Expected non-nil success argument 'theValue' for nil error")
        }
        continuation.resume(returning: theValue)
      }
    }
  }
}
```

## Example 2 (convert to async/await)
### Base Code
```swift
func test(completion: (String?, Error?) -> Void) {
  simpleErr() { (res, err) in
    completion(res, err)
  }
}
```
### Old Refactoring Result
```swift
func test() async throws -> String {
  let res = try await simpleErr()
  throw <#err#>
}
```
### New Refactoring Result
```swift
func test() async throws -> String {
  let res = try await simpleErr()
  return res
}
```

hamishknight reviewed (commented):
Nice!

ahoppen commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6576,7 +6609,27 @@ class AsyncConverter : private SourceEntityWalker {
+        }
+      }
+
+      if (ErrorIsPlaceholered) {
+        // The error parameter's decleration doesn't exist anymore. This is
+        // probably because it's been handled by a 'try await'. Since the
+        // original call was ambiguous, the result case seems more reasonable
+        // now.
+        Exprs = TopHandler.extractResultArgs(CE, /*ReturnErrorArgsIfAmbiguous=*/false);
+      }
```
That does sound like a reasonable and simpler idea. 👍

hamishknight replied:
Do you think it's worth adding as a test case?

ahoppen replied:
We are actually by now converting this to code without placeholders because of all the various changes 🎉 Still, I added a test case because you can never have enough test cases.

```swift
func testReturnHandling4() async throws -> String {
  do {
    let str = try await simpleErr(arg: "xxx")
    return str
  } catch let err {
    print("some error stuff")
    throw err
  }
}
```

hamishknight replied:
Thanks!

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
To align the generated code with "Add Async Wrapper", I changed it to use `guard let else { fatalError }` to check that all success parameters are non-nil when the error is nil. 

I also ran `git-clang-format` for #38193 because I forgot it there…

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6640,17 +6767,80 @@ class AsyncConverter : private SourceEntityWalker {
+           << tok::equal << ' ';
+        convertNode(Arg, /*StartOverride=*/CE->getArgumentLabelLoc(I),
+                    /*ConvertCalls=*/false);
+        if (auto CE = dyn_cast<CallExpr>(Arg)) {
+          if (CE->hasTrailingClosure()) {
+            // If the argument is a call with trailing closure, the generated
+            // guard statement does not compile.
+            // e.g. 'guard let result1 = value.map { $0 + 1 } else { ... }'
+            // doesn't compile.
+            // Adding a '.self' at the end makes the code compile (although it
+            // will still issue a warning about a trailing closure use inside a
+            // guard condition).
+            OS << tok::period << tok::kw_self;
+          }
+        }
```
Is it worth also handling this case for the error argument that's used in the `if let`? Maybe it should be part of `convertNode`?

ahoppen replied:
For some reasons, trailing closures are fine in `if let` but not in `guard let`. And since the `.self` isn’t particularly nice if it can be avoided, I’d prefer to leave it as-is.

hamishknight replied:
Ah interesting! If anything I would have expected it be the other way around because of the `else` keyword. I agree it's best to leave this as-is then.

ahoppen replied:
So would I but 🤷

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6640,17 +6767,80 @@ class AsyncConverter : private SourceEntityWalker {
+        ArgNames->push_back(ArgName);
+        OS << tok::kw_guard << ' ' << tok::kw_let << ' ' << ArgName << ' '
+           << tok::equal << ' ';
+        convertNode(Arg, /*StartOverride=*/CE->getArgumentLabelLoc(I),
```
The `I` used by the loop starts on 1, but this usage expects it to start on 0. I don't *think* it currently causes issues here as the handler being called should be a function value that cannot have argument labels (and the presence of the error param means we won't run off the end), but could this still be corrected in case argument labels are supported in the future?

ahoppen replied:
Changed `I` to start at 0, which matches better with tuple indexing anyway.

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6847,15 +7038,18 @@ class AsyncConverter : private SourceEntityWalker {
-    preparePlaceholdersAndUnwraps(HandlerDesc, SuccessParams, ErrParam,
-                                  /*Success=*/true);
+    {
+      llvm::SaveAndRestore<bool> InSuccesBranch(ConvertingSuccessBlocks, true);
```
What do you think about moving this into `convertNodes`, and having `convertNodes` take a bool param for success or failure? That way if we add more uses of `convertNodes` in the future we won't forget to update this state.

ahoppen replied:
Good idea, because I only now realized that we weren’t setting `ConvertingSuccessBlocks` to `false` when converting an error block nested in a success block. So the suggestion already caught a potential bug.

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6601,6 +6640,34 @@ class AsyncConverter : private SourceEntityWalker {
+        //   completion(res!, nil)
+        // which correctly yields
+        //   return res
+        // Syntesize the force unwrap so that we get the expected results.
```
Typo: Syntesize

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6609,20 +6676,80 @@ class AsyncConverter : private SourceEntityWalker {
     assert(Scopes.back().isWrappedInContination());
 
-    ArrayRef<Expr *> Args;
+    std::vector<Expr *> Args;
```
Out of curiosity, why did this need to become a vector?

ahoppen replied:
That’s because actually before this might result in unsafe memory access. 🙈 

In the params case, `Args` references into `Exprs`, which goes out of scope before `Args`. Making it a `std::vector` makes sure that all `Expr` pointers stay alive. I already cherry-picked this fix to the 5.5 PR #38256.

hamishknight replied:
Ah, I missed the fact that `HandlerResult` can own the elements in the single element case, good catch!

hamishknight reviewed (commented):
A few more comments, but otherwise LGTM!

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6640,17 +6769,80 @@ class AsyncConverter : private SourceEntityWalker {
+          } else {
+            // We are returning a tuple. Name the result elements 'result' +
+            // index in tuple.
+            ArgName = createUniqueName("result" + std::to_string(I)).str();
```
Should this be changed to be `I + 1` after the indexing change? IMO that's a little more natural for naming. It seems we also don't have a test case that exercises this, would it be worth adding one?

ahoppen replied:
That’s why I had `I` be `1`-based initially. But then I realized that if you have a tuple `let result: (Int, Int)`, you also access it `0`-based (e.g. `result.0`), so I thought that it’s also natural to use `0`-based indexing here.

ahoppen replied:
And I thought I added a test case for this but apparently but apparently I didn’t. Did so now.

hamishknight replied:
Hmm, yeah that's a fair point, I don't feel too strongly about it either way

ahoppen replied:
Let’s leave it `0`-based then. I think it looks rather pretty in the new test case
```swift
continuation.resume(returning: (result0, result1))
```

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on test/refactoring/ConvertAsync/convert_function.swift:
```diff
@@ -345,4 +366,26 @@ func withImplicitReturn(completionHandler: (String) -> Void) {
+// RELAY-ERROR-AND-RESULT:      func relayErrorAndResult() async throws -> String {
+// RELAY-ERROR-AND-RESULT-NEXT:   let res = try await simpleErr(arg: "test")
+// RELAY-ERROR-AND-RESULT-NEXT:   return res
+// RELAY-ERROR-AND-RESULT-NEXT: }
```
Add a test for non-nil error here? Ie.
```
simple { res in
  completion(res, MyError())
}
```

ahoppen replied:
Added one and realized that we were actually dropping the error there (because we had an ambiguous call inside the success block). I changed the heuristic to 
- only interpret an ambiguous call as a success call if it isn the success branch and the error has been placeholdered
- add `assert`s that the success arguments are `nil` if we continue to interpret the ambiguous call as an error call

bnbarham commented on test/refactoring/ConvertAsync/convert_to_continuation.swift:
```diff
@@ -101,11 +102,139 @@ func testThrowingContinuationRelayingErrorAndResult(completionHandler: (Int?, Er
+// RUN: %refactor -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=THROWING-CONTINUATION-ALWAYS-RETURNING-ERROR-AND-RESULT %s
+func testAlwaysReturnBothResultAndCompletionHandler(completionHandler: (Int?, Error?) -> Void) {
+  withoutAsyncAlternativeBecauseOfMismatchedCompletionHandlerName { theValue in
+    completionHandler(theValue, MyError())
```
What do you think about just turning this into eg. `continuation.resume(throwing: MyError())` without any of the ifs/guards? Ie. If the error value is non-optional, always use it. Or we could go the other way, ie. if the *success* value is non-optional, always use it. I think the error one makes more sense though.

And in that case I'd probably not worry about adding the assert for any of the cases + only add the guard let for the success when they are optional.

ahoppen replied:
Good idea. I’m still in favor of keeping the `assert`s because it makes assumptions about our generated code explicit so that we don’t silently drop results.

ahoppen commented:
Thanks for you feedback @bnbarham and @hamishknight. I’ve
- Changed the heuristic to interpret an ambiguous completion call as a success call only if we are in the success branch and the error has been placeholdered
- Added `assert`s verifying that all success parameters are `nil` if we are interpreting an ambiguous completion call as an error call
- Added some more test cases

Here’s a diff between the last and the latest version: https://github.com/ahoppen/swift/compare/80506eb7071b38e302a493520cc11f2b12671890..0c6b9d5790e4382d6cea5ac5b956adc2c62ef7f6

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on test/refactoring/ConvertAsync/convert_function.swift:
```diff
@@ -342,4 +363,38 @@ func withImplicitReturn(completionHandler: (String) -> Void) {
+    completion(res, CustomError.Bad)
+  }
+}
+// ALWAYS-THROW-ERROR-AND-RETURN-RESULT:      func alwaysThrowErrorAndReturnResult() async throws -> String {
+// ALWAYS-THROW-ERROR-AND-RETURN-RESULT-NEXT:   let res = await simple()
+// ALWAYS-THROW-ERROR-AND-RETURN-RESULT-NEXT:   assert(res == nil, "The code this function was generated from passed a result and an error argument to the completion handler. Since this function just throws the error, it is assumed that the result was nil.")
+// ALWAYS-THROW-ERROR-AND-RETURN-RESULT-NEXT:   throw CustomError.Bad
+// ALWAYS-THROW-ERROR-AND-RETURN-RESULT-NEXT: }
```
Could you make sure we have test cases for a `(String?, Error?) -> Void` completion handler where:
1. Result is non-optional and error is optional
2. Result is optional and error is non-optional
3. Both are optional
4. Both are non-optional
Ideally with simple closure args/aliases as well as unknown expressions (ie. a random call/error/whatever).

I don't see a test for (nil, nil) anywhere so if you could add that one that'd be great :).

Finally, in regards to this test specifically, I'd prefer a placeholder and comment rather than the assert (since the compiler error could be confusing for people). Suggested wording could be:
```
// The original completion handler call passed a non-optional error. An additional result cannot be represented by the async function and just a throws has been generated instead. Note that this drops the result.
```
Note that if the error expression *is* optional we could save it to a variable and then do an `if` instead (ie. the same as the continuation refactoring).

hamishknight replied:
Yeah I also think a placeholder + comment would be a little clearer to the user in that case (though I hope it doesn't come up too often)

ahoppen committed:
[Async Refactoring] Convert data structure storing continuation resume args to std::vector

The `std::vector` keeps its elements alive. Currently, we might be accessing uninitalized memory if the call of `callArgs(CE)` (which owns its elements in case there is a single element) goes out of scope before `Args`.

ahoppen committed:
[Async Refactoring] Split ambiguous (error + success) completion handler calls into success and error case

Previously, when a completion handler call had arguments for both the success and error parameters, we were always interpreting it as an error call. In case the error argument was an `Optional`, this could cause us to generate code that didn't compile, because we `throw`ed the `Error?` or passed it to `continuation.resume(throwing)`, both of which didn't work.

We now generate an `if let` statement that checks if the error is `nil`. If it is, the error is thrown. Otherwise, we interpret the call as a success call and return the result.

- Example 1 (convert to continuation)
-- Base Code
```swift
func test(completionHandler: (Int?, Error?) -> Void) {
  withoutAsyncAlternativeThrowing { (theValue, theError) in
    completionHandler(theValue, theError)
  }
}
```
-- Old Refactoring Result
```swift
func test() async throws -> Int {
  return try await withCheckedThrowingContinuation { continuation in
    withoutAsyncAlternativeThrowing { (theValue, theError) in
      continuation.resume(throwing: theError) // error: Argument type 'Error?' does not conform to expected type 'Error'
    }
  }
}
-- New Refactoring Result
```swift
func testThrowingContinuationRelayingErrorAndResult() async throws -> Int {
  return try await withCheckedThrowingContinuation { continuation in
    withoutAsyncAlternativeThrowing { (theValue, theError) in
      if let error = theError {
        continuation.resume(throwing: error)
      } else {
        guard let theValue = theValue else {
          fatalError("Expected non-nil success argument 'theValue' for nil error")
        }
        continuation.resume(returning: theValue)
      }
    }
  }
}
```

- Example 2 (convert to async/await)
-- Base Code
```swift
func test(completion: (String?, Error?) -> Void) {
  simpleErr() { (res, err) in
    completion(res, err)
  }
}
```
-- Old Refactoring Result
```swift
func test() async throws -> String {
  let res = try await simpleErr()
  throw <#err#>
}
```
-- New Refactoring Result
```swift
func test() async throws -> String {
  let res = try await simpleErr()
  return res
}
```

ahoppen committed:
[Async Refactoring] Remove assertions for nil result values in async wrapper

Since we aren’t checking that all result arguments are `nil` if there was an `error` in all the other refactorings, also remove the assertions in the async wrapper.

ahoppen committed:
[Async Refactoring] Rename 'success param/argument' to 'result' in `fatalError` messages

The 'success param/argument' is a terminology we use internally in the refactoring and not one we want to present to the user. Just name it 'result' instead.

ahoppen commented:
OK, so I’ve completely rewritten the entire implementation. The logic for determining how to interpret a completion handler call is now shared between the `async`/`await` and continuation refactoring in [`convertHandlerCall`](https://github.com/apple/swift/pull/38227/files#diff-527d7f50231797f1eb68d5efeaaa33de5e5f324aa303c1cf20dd3f7d6d879f04R6670-R6726). IMHO it has a fairly concise description how the interpretation works instead of the hot-patching I had earlier.

The general rule is now: If there is a non-`nil` error, then we throw it. If there are any non-nil results, we ignore them (without any assertions or whatsoever). If the error is `nil`, we return the result. If the result is `nil`, this results in a compile or runtime error. IIRC this exactly matches what the clang importer is doing.

Consequently, I also removed the assertions for the async wrapper I added in #38259.

ahoppen commented:
@swift-ci Please smoke test

hamishknight reviewed (approved):
LGTM!

bnbarham commented on test/refactoring/ConvertAsync/convert_function.swift:
```diff
@@ -342,4 +366,210 @@ func withImplicitReturn(completionHandler: (String) -> Void) {
+}
+// NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR:      func nonOptionalResultAndNonOptionalError() async throws -> String {
+// NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR-NEXT:   throw CustomError.Bad
+// NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR-NEXT: }
```
Thanks for adding all these tests! Much better coverage now IMO.

bnbarham replied:
Just thought of another test as well - simple arguments wrapped in parentheses.

ahoppen replied:
Since the logic is exactly the same as for the continuation refactoring, the parenthesis case is already covered by #38261. Added a test case in `convert_function` anyway.

bnbarham commented on test/refactoring/ConvertAsync/convert_to_continuation.swift:
```diff
@@ -341,6 +448,100 @@ func testResultFromValueAndError(completionHandler: (Result<Int, Error>) -> Void
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       guard let theResult1 = theResult else {
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:         fatalError("Expected non-nil result 'theResult1' for nil error")
```
The message here is a bit odd since error is completely ignored... Not sure it's worth changing in this case though.

ahoppen replied:
What do you think about changing *all* messages to one of
- `Expected non-nil result '...' in the success case`
- `Expected non-nil result '...' in the non-error case`
- `Expected non-nil result '...' if no error is thrown`

I think even in the 
```swift
if let error = error {
  continuation.resume(throwing: error)
} else {
  guard let first1 = first else {
    fatalError("Expected non-nil result 'first1' for nil error")
  }
  continuation.resume(returning: first1)
}
```
it’s not entirely clear what `nil error` refers to.

ahoppen replied:
Thinking about it, I went with the `Expected non-nil result '...' in the non-error case` message. 

I also added a special case to omit the result name if it is "result". I.e. instead of `Expected non-nil result 'result' in non-error case`, we now emit `Expected non-nil result in non-error case`. If the result variable has any other name, we still emit it.

bnbarham commented on test/refactoring/ConvertAsync/convert_to_continuation.swift:
```diff
@@ -341,6 +448,100 @@ func testResultFromValueAndError(completionHandler: (Result<Int, Error>) -> Void
+// PASS-SAME-RESULT-TWICE:      func testPassSameResultTwice() async throws -> (Int, Int) {
+// PASS-SAME-RESULT-TWICE-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
+// PASS-SAME-RESULT-TWICE-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
+// PASS-SAME-RESULT-TWICE-NEXT:       guard let theResult1 = theResult else {
```
How hard would this be to avoid? If not much effort it'd be good to fix, but probably an uncommon case so I'm fine with leaving.

ahoppen replied:
Probably not too much if we only worry about the *exact same* argument being passed to the completion handler twice. But I think there’s very little benefit in it (who passes the same argument to the completion handler twice but me?) and I’m more worried that the logic will introduce bugs in the common cases.

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -6588,14 +6614,137 @@ class AsyncConverter : private SourceEntityWalker {
+        // interpret it as an error call.
+        AddConvertedHandlerCall(Result);
+      } else {
+        // The call was truely amiguous. Add an
```
truely -> truly
amiguous -> ambiguous

bnbarham reviewed (approved):
Had a few small comments but otherwise the new changes LGTM :)!

ahoppen committed:
[Async Refactoring] Improve fatalError message to unwrap result arguments in continuation

In the previous `fatalError` message `Expected non-nil result 'result' for nil error`, it wasn’t exactly clear wht the `error` referred to. In some cases, when the error was ignored, there wasn’t even an `error` variable in the refactored code.

The new `Expected non-nil result '...' in the non-error case` is a little more generic and also fits if an error is ignored.

ahoppen commented:
For easier review, here’s the diff between the last and current version

<details>
<summary>git diff 8fc37f8..cd31fa0</summary>

```diff
diff --git a/lib/IDE/Refactoring.cpp b/lib/IDE/Refactoring.cpp
index 388b2504c8b..72c2f4b1c31 100644
--- a/lib/IDE/Refactoring.cpp
+++ b/lib/IDE/Refactoring.cpp
@@ -6700,7 +6700,7 @@ private:
         // interpret it as an error call.
         AddConvertedHandlerCall(Result);
       } else {
-        // The call was truely amiguous. Add an
+        // The call was truly ambiguous. Add an
         // if let error = <convert error arg> {
         //   throw error // or equivalent
         // } else {
@@ -6902,7 +6902,11 @@ private:
         }
         OS << ' ' << tok::kw_else << ' ' << tok::l_brace << '\n';
         OS << "fatalError" << tok::l_paren;
-        OS << "\"Expected non-nil result '" << ArgName << "' for nil error\"";
+        OS << "\"Expected non-nil result ";
+        if (ArgName.str() != "result") {
+          OS << "'" << ArgName << "' ";
+        }
+        OS << "in the non-error case\"";
         OS << tok::r_paren << '\n';
         OS << tok::r_brace << '\n';
       }
diff --git a/test/refactoring/ConvertAsync/convert_function.swift b/test/refactoring/ConvertAsync/convert_function.swift
index 303fbee102d..90b816a4d4e 100644
--- a/test/refactoring/ConvertAsync/convert_function.swift
+++ b/test/refactoring/ConvertAsync/convert_function.swift
@@ -572,4 +572,28 @@ func nonOptionalResultAndNonOptionalError(completion: (String?, Error?) -> Void)
 }
 // NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR:      func nonOptionalResultAndNonOptionalError() async throws -> String {
 // NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR-NEXT:   throw CustomError.Bad
-// NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR-NEXT: }
\ No newline at end of file
+// NON-OPTIONAL-RESULT-AND-NON-OPTIONAL-ERROR-NEXT: }
+
+// RUN: %refactor -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=WRAP-COMPLETION-CALL-IN-PARENS %s
+func wrapCompletionCallInParenthesis(completion: (String?, Error?) -> Void) {
+  simpleErr(arg: "test") { (res, err) in
+    (completion(res, err))
+  }
+}
+// WRAP-COMPLETION-CALL-IN-PARENS:      func wrapCompletionCallInParenthesis() async throws -> String {
+// WRAP-COMPLETION-CALL-IN-PARENS-NEXT:   let res = try await simpleErr(arg: "test")
+// WRAP-COMPLETION-CALL-IN-PARENS-NEXT:   return res
+// WRAP-COMPLETION-CALL-IN-PARENS-NEXT: }
+
+// RUN: %refactor -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=TWO-COMPLETION-HANDLER-CALLS %s
+func twoCompletionHandlerCalls(completion: (String?, Error?) -> Void) {
+  simpleErr(arg: "test") { (res, err) in
+    completion(res, err)
+    completion(res, err)
+  }
+}
+// TWO-COMPLETION-HANDLER-CALLS:      func twoCompletionHandlerCalls() async throws -> String {
+// TWO-COMPLETION-HANDLER-CALLS-NEXT:   let res = try await simpleErr(arg: "test")
+// TWO-COMPLETION-HANDLER-CALLS-NEXT:   return res
+// TWO-COMPLETION-HANDLER-CALLS-NEXT:   return res
+// TWO-COMPLETION-HANDLER-CALLS-NEXT: }
diff --git a/test/refactoring/ConvertAsync/convert_to_continuation.swift b/test/refactoring/ConvertAsync/convert_to_continuation.swift
index 7bc517c18f9..d03d8aecc5a 100644
--- a/test/refactoring/ConvertAsync/convert_to_continuation.swift
+++ b/test/refactoring/ConvertAsync/convert_to_continuation.swift
@@ -120,7 +120,7 @@ func testThrowingContinuationRelayingErrorAndResult(completionHandler: (Int?, Er
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:         continuation.resume(throwing: error)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:       } else {
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:         guard let theValue1 = theValue else {
-// THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:           fatalError("Expected non-nil result 'theValue1' for nil error")
+// THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:           fatalError("Expected non-nil result 'theValue1' in the non-error case")
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:         }
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:         continuation.resume(returning: theValue1)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-RESULT-NEXT:       }
@@ -141,7 +141,7 @@ func testThrowingContinuationRelayingErrorAndComplexResult(completionHandler: (I
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:         continuation.resume(throwing: error)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:       } else {
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:         guard let result = theValue.map({ $0 + 1 }) else {
-// THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:           fatalError("Expected non-nil result 'result' for nil error")
+// THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:           fatalError("Expected non-nil result in the non-error case")
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:         }
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:         continuation.resume(returning: result)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-NEXT:       }
@@ -162,10 +162,10 @@ func testThrowingContinuationRelayingErrorAndTwoComplexResults(completionHandler
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:         continuation.resume(throwing: error)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:       } else {
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:         guard let result0 = theValue.map({ $0 + 1 }) else {
-// THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:           fatalError("Expected non-nil result 'result0' for nil error")
+// THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:           fatalError("Expected non-nil result 'result0' in the non-error case")
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:         }
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:         guard let result1 = theValue.map({ $0 + 2 }) else {
-// THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:           fatalError("Expected non-nil result 'result1' for nil error")
+// THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:           fatalError("Expected non-nil result 'result1' in the non-error case")
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:         }
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:         continuation.resume(returning: (result0, result1))
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-TWO-COMPLEX-RESULTS-NEXT:       }
@@ -186,7 +186,7 @@ func testThrowingContinuationRelayingErrorAndComplexResultWithTrailingClosure(co
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:         continuation.resume(throwing: error)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:       } else {
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:         guard let result = theValue.map { $0 + 1 }.self else {
-// THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:           fatalError("Expected non-nil result 'result' for nil error")
+// THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:           fatalError("Expected non-nil result in the non-error case")
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:         }
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:         continuation.resume(returning: result)
 // THROWING-CONTINUATION-RELAYING-ERROR-AND-COMPLEX-RESULT-WITH-TRAILING-CLOSURE-NEXT:       }
@@ -461,10 +461,10 @@ func testMultipleReturnValuesAndError(completion: (Int?, String?, Error?) -> Voi
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:         continuation.resume(throwing: error)
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:       } else {
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:         guard let first1 = first else {
-// MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:           fatalError("Expected non-nil result 'first1' for nil error")
+// MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:           fatalError("Expected non-nil result 'first1' in the non-error case")
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:         }
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:         guard let second1 = second else {
-// MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:           fatalError("Expected non-nil result 'second1' for nil error")
+// MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:           fatalError("Expected non-nil result 'second1' in the non-error case")
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:         }
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:         continuation.resume(returning: (first1, second1))
 // MULTIPLE-RETURN-VALUES-AND-ERROR-NEXT:       }
@@ -496,7 +496,7 @@ func testMixedOptionalAnNonOptionaResults(completion: (Int?, String?, Error?) ->
 // MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
 // MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
 // MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       guard let theResult1 = theResult else {
-// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:         fatalError("Expected non-nil result 'theResult1' for nil error")
+// MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:         fatalError("Expected non-nil result 'theResult1' in the non-error case")
 // MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       }
 // MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:       continuation.resume(returning: (theResult1, "hi"))
 // MIXED-OPTIONAL-AND-NON-OPTIONAL-RESULT-NEXT:     }
@@ -514,7 +514,7 @@ func testUseOptionalResultValueAfterCompletionHandlerCall(completion: (Int?, Str
 // USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
 // USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
 // USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       guard let theResult1 = theResult else {
-// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:         fatalError("Expected non-nil result 'theResult1' for nil error")
+// USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:         fatalError("Expected non-nil result 'theResult1' in the non-error case")
 // USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       }
 // USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       continuation.resume(returning: (theResult1, "hi"))
 // USE-OPTIONAL-RESULT-AFTER-COMPLETION-HANDLER-CALL-NEXT:       print(theResult.map { $0 + 1 } as Any)
@@ -533,10 +533,10 @@ func testPassSameResultTwice(completion: (Int?, Int?, Error?) -> Void) {
 // PASS-SAME-RESULT-TWICE-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
 // PASS-SAME-RESULT-TWICE-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
 // PASS-SAME-RESULT-TWICE-NEXT:       guard let theResult1 = theResult else {
-// PASS-SAME-RESULT-TWICE-NEXT:         fatalError("Expected non-nil result 'theResult1' for nil error")
+// PASS-SAME-RESULT-TWICE-NEXT:         fatalError("Expected non-nil result 'theResult1' in the non-error case")
 // PASS-SAME-RESULT-TWICE-NEXT:       }
 // PASS-SAME-RESULT-TWICE-NEXT:       guard let theResult2 = theResult else {
-// PASS-SAME-RESULT-TWICE-NEXT:         fatalError("Expected non-nil result 'theResult2' for nil error")
+// PASS-SAME-RESULT-TWICE-NEXT:         fatalError("Expected non-nil result 'theResult2' in the non-error case")
 // PASS-SAME-RESULT-TWICE-NEXT:       }
 // PASS-SAME-RESULT-TWICE-NEXT:       continuation.resume(returning: (theResult1, theResult2))
 // PASS-SAME-RESULT-TWICE-NEXT:     }
@@ -544,6 +544,60 @@ func testPassSameResultTwice(completion: (Int?, Int?, Error?) -> Void) {
 // PASS-SAME-RESULT-TWICE-NEXT: }
 
 
+// RUN: %refactor-check-compiles -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL %s
+func testUseResultAfterAmbiguousCompletionHandlerCall(completion: (Int?, Error?) -> Void) {
+  withoutAsyncAlternativeThrowing { (theResult, error) in
+    completion(theResult, error)
+    print(theResult as Any)
+  }
+}
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL:      func testUseResultAfterAmbiguousCompletionHandlerCall() async throws -> Int {
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:       if let error = error {
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:         continuation.resume(throwing: error)
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:       } else {
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:         guard let theResult1 = theResult else {
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:           fatalError("Expected non-nil result 'theResult1' in the non-error case")
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:         }
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:         continuation.resume(returning: theResult1)
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:       }
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:       print(theResult as Any)
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:     }
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT:   }
+// USE-RESULT-AFTER-AMBIGUOUS-HANLDER-CALL-NEXT: }
+
+// RUN: %refactor-check-compiles -convert-to-async -dump-text -source-filename %s -pos=%(line+1):1 | %FileCheck -check-prefix=TWO-COMPLEITON-HANDLER-CALLS %s
+func testTwoCompletionHandlerCalls(completion: (Int?, Error?) -> Void) {
+  withoutAsyncAlternativeThrowing { (theResult, error) in
+    completion(theResult, error)
+    completion(theResult, error)
+  }
+}
+// TWO-COMPLEITON-HANDLER-CALLS:      func testTwoCompletionHandlerCalls() async throws -> Int {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:   return try await withCheckedThrowingContinuation { continuation in 
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:     withoutAsyncAlternativeThrowing { (theResult, error) in
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:       if let error = error {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         continuation.resume(throwing: error)
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:       } else {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         guard let theResult1 = theResult else {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:           fatalError("Expected non-nil result 'theResult1' in the non-error case")
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         }
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         continuation.resume(returning: theResult1)
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:       }
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:       if let error = error {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         continuation.resume(throwing: error)
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:       } else {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         guard let theResult2 = theResult else {
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:           fatalError("Expected non-nil result 'theResult2' in the non-error case")
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         }
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:         continuation.resume(returning: theResult2)
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:       }
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:     }
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT:   }
+// TWO-COMPLEITON-HANDLER-CALLS-NEXT: }
+
+
 // Reduced version of https://twitter.com/peterfriese/status/1397835146133479428
 
 class DataTask {
```
</details>

ahoppen commented:
@swift-ci Please smoke test
